### ISSUE TITLE
403 Forbidden after adding custom route

### ISSUE THREAD

**Issue by genu** [strapi/strapi#3423]
**Describe the bug**
I added a custom route on a model and enabled the route in user-permissions.  (Public) Accessing the route via browser or XHR results in `403 Forbidden` response

**Steps to reproduce the behavior**
1. Add a new custom route
2. `/admin/plugins/users-permissions/roles`
3. Click on `Public`
4. Check the box next to the newly created route
5. Try to access the route in the browser. 

**Expected behavior**
I expect the route to at least be hit, but code never reaches the controller method

**Code snippets**
````
{
      "method": "POST",
      "path": "/validate",
      "handler": "Product.validate",
      "config": {
        "policies": []
      }
    }
````
**System**
- Node.js version: v10.16.0
- NPM version: 6.9.0
- Strapi version: 3.0.0-beta.3
- Database: postgres
- Operating system: OSx

**Comment by alexandrebodin** [comment]
Hi @genu,

This works fine locally. can you share the query and the headers / params you are sending please ?

**Comment by genu** [comment]
I'm not sending any special headers just standard:
````
GET /children/needingSponsorship HTTP/1.1
Host: localhost:1337
Connection: close
User-Agent: Paw/3.1.8 (Macintosh; OS X/10.14.5) GCDHTTPRequest
````

Also I'm not doing a query, I'm just returning a basic response

controller:
````
validate: async (ctx, next) => {
    console.log('test')
  }
```

the console statement is never reached.

Also note, the standard, builin CRUD functions that ship with Strapi work correctly.

**Comment by genu** [comment]
Sorry, typo on my part. Updated the the snipped. Route is a `GET`

**Comment by alexandrebodin** [comment]
This works fine doing the exact same routes.

Can you share a git repo with that error reproduced ?

**Comment by genu** [comment]
hmm. Let me try to setup up repo and share.

**Comment by genu** [comment]
ok, after further investigation I think I found the issue.

It appears that the ORDER of the routes matter. So If I put the custom `GET` route more to the top of the `routes.json` it work properly. But if I put it at the end of the array, will produce the 403.

Did I miss something in the docs related to the route ordering?

**Comment by alexandrebodin** [comment]
This will to that if you have multiple times the route /validate otherwise it doesn't matter

**Comment by alexandrebodin** [comment]
I think there is a conflict with the name validate. Try changing it's name and let me know

**Comment by genu** [comment]
ok, so there is a conflict.

The conflict is:
````
{
      "method": "GET",
      "path": "/products/:id",
      "handler": "Product.findOne",
      "config": {
        "policies": []
      }
},
````

will take priority over:
````
    {
      "method": "GET",
      "path": "/products/validate",
      "handler": "Product.validate",
      "config": {
        "policies": []
      }
    }
````

I believe in previous versions of strapi this behavior was different.

So, i suppose if I want to make this type of non-crud `GET` I'd need to make sure to place it properly in the `routes.json`?

**Comment by alexandrebodin** [comment]
Yes this behaviour hasn't been changed in the beta. You will need to make sure your routes aren't conflicting with each others :)

**Comment by JiboStore** [comment]
Sorry for commenting on closed issue but @genu 

```
And the 403 is a result of Product.findOne not enabled in permissions.
```

how do you enable the function `Product.findOne` I am facing similar issue, I wrote a `Product.findOne` myself but the route is always forbidden even though I have no other similar route

**Comment by ghonchesefidi** [comment]
As @genu mentioned the order of new routes is mather. So if you are adding a new route to an existing routes.js file, you should add it right after routes at the top of others.

**Comment by derrickmehaffy** [comment]
> As @genu mentioned the order of new routes is mather. So if you are adding a new route to an existing routes.js file, you should add it right after routes at the top of others.

Yup the key point is routes are match via regex so if you are adding a custom route similar to findOne but with something other than an ID it will match the default `findOne` route (the regex matches the other value as an ID)

So if you are implementing a custom findOne for a slug you need to place it above, noting that the default findOne may never be called. In a case where you need both, you probably do need to expand the path entirely say for example:

```json
    {
      "method": "GET",
      "path": "/clients/:id",
      "handler": "client.findOne",
      "config": {
        "policies": []
      }
    },
    {
      "method": "GET",
      "path": "/clients/slug/:slug",
      "handler": "client.findOneBySlug",
      "config": {
        "policies": []
      }
    },
```

**Comment by rullymartanto** [comment]
I still face the same issues...

`{
  "routes": [
    {
      "method": "POST",
      "path": "/contacts/captcha",
      "handler": "contact.captcha",
      "config": {
        "policies": []
      }
    },
    {
      "method": "GET",
      "path": "/contacts",
      "handler": "contact.find",
      "config": {
        "policies": []
      }
    },
   {
      "method": "POST",
      "path": "/contacts",
      "handler": "contact.create",
      "config": {
        "policies": []
      }
    },`

Also I have try 
`{
  "routes": [
    {
      "method": "POST",
      "path": "/contacts/captcha/validate",
      "handler": "contact.captcha",
      "config": {
        "policies": []
      }
    },
    {
      "method": "GET",
      "path": "/contacts",
      "handler": "contact.find",
      "config": {
        "policies": []
      }
    },
   {
      "method": "POST",
      "path": "/contacts",
      "handler": "contact.create",
      "config": {
        "policies": []
      }
    },`

**Comment by Marcel-Tronco** [comment]
I have a similar issue. A custom POST route with only one entry, so ordering should not be the issue. I get a 403.
Here's  the routes.json:
```
{
  "routes": [
    {
      "method": "POST",
      "path": "/deploy",
      "handler": "Deploy.index",
      "config": {
        "policies": []
      }
    }
  ]
}
```

**Comment by derrickmehaffy** [comment]
> I still face the same issues...
> 
> `{ "routes": [ { "method": "POST", "path": "/contacts/captcha", "handler": "contact.captcha", "config": { "policies": [] } }, { "method": "GET", "path": "/contacts", "handler": "contact.find", "config": { "policies": [] } }, { "method": "POST", "path": "/contacts", "handler": "contact.create", "config": { "policies": [] } },`
> 
> Also I have try
> `{ "routes": [ { "method": "POST", "path": "/contacts/captcha/validate", "handler": "contact.captcha", "config": { "policies": [] } }, { "method": "GET", "path": "/contacts", "handler": "contact.find", "config": { "policies": [] } }, { "method": "POST", "path": "/contacts", "handler": "contact.create", "config": { "policies": [] } },`

Did you set the permissions?

**Comment by rullymartanto** [comment]
yes it's work. wrong set place of roles and permissions.
Thanks
